event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java

```diff
@@ -20,20 +20,53 @@
 package org.apache.james.eventsourcing.eventstore.cassandra;
 
 import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
 import java.util.Set;
 
 import javax.inject.Inject;
+import javax.inject.Named;
 
 import org.apache.james.eventsourcing.Event;
 import org.apache.james.eventsourcing.eventstore.cassandra.dto.EventDTO;
 import org.apache.james.eventsourcing.eventstore.cassandra.dto.EventDTOModule;
+import org.apache.james.json.DTOModule;
 import org.apache.james.json.JsonGenericSerializer;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
+import com.github.steveash.guavate.Guavate;
 import com.google.common.collect.ImmutableSet;
 
 public class JsonEventSerializer {
 
+    public static RequireNestedConfiguration forModules(Set<? extends EventDTOModule<?, ?>> modules) {
+        return nestedTypesModules -> {
+            ImmutableSet<EventDTOModule<?, ?>> dtoModules = ImmutableSet.copyOf(modules);
+            return new JsonEventSerializer(dtoModules, ImmutableSet.copyOf(nestedTypesModules));
+        };
+    }
+
+    @SafeVarargs
+    public static RequireNestedConfiguration forModules(EventDTOModule<?, ?>... modules) {
+        return forModules(ImmutableSet.copyOf(modules));
+    }
+
+    public interface RequireNestedConfiguration {
```

Comment:
Wouldn't it be clearer if it was the responsibility of the module to provide the module it requires to handle the nested object of its DomainObject / DTO ?

Reply:
what do you mean?

Reply:
I thought that the dependency between modules could be expressed in the module.
With each module containing a list of it's direct child module.
This way we have the tree of dependency in the module and we don't have to figure out every subdependency which can be defined at a deep sub-level, when we define the serializer.

Maybe i'm wrong but it seems easy to make errors by forgetting nested modules when declaring the serializer.

Reply:
That's an interesting suggestion, I'll give it a try.

Reply:
After some thoughts, it just can't work.
Let's take the distributed task manager example: we built a system that can distribute task between nodes. The system doesn't know much about the tasks that exist in a given system.
However, it needs to be able to serialize/deserialize them as part of its own types (Created, Started, etc).
To achieve that, we defined `TaskSerializationModuleFactory` type: we declare a module that will later receive the information about the types it will have to embed.
Said another way; we delayed the dependency tree creation at injection time.
And we have no other choice because we build the set of available tasks once every modules are declared.

Reply:
Ok, i think i understand.